### PR TITLE
Feature: moved CORE done by Facu to clean PR

### LINKED ISSUE #92: CORE - Calcular aportes y deudas
Puro. Deterministico. SOLIDO.

- [x] Pensarlo - 2
- [x] Documentarlo - 1
- [x] Hacerlo - 2
- [x] Testearlo - 1

Routes
- R (any) Leer Aportes
- R (any) Leer Deudas

Image

### PR DESCRIPTION
**Pull request by MateoDi9z**
This pull request introduces a new core balance calculation engine for the project, centered around a `BalancesMap` structure and a deterministic `core` function that reconstructs group and user balances from transaction and expense histories. It also adds a comprehensive test suite to ensure correctness and invariants of the balance logic, including edge cases and integration scenarios.

Core balance logic:

* Implements the `BalancesMap` struct in `balances_map.rs`, providing methods for adding balances to users, distributing withdrawals or expenses across all users, and ensuring group/user balance invariants with robust error handling.
* Adds the `core` function in `core.rs`, which deterministically computes current balances from a list of users, transactions, and expenses, enforcing strict integrity checks and handling various transaction types.

Testing and modularization:

* Introduces a thorough test suite covering transactions, expenses, and integration scenarios, validating invariants and error cases (e.g., unknown users, insufficient funds), with test helpers for generating test data. [[1]](diffhunk://#diff-d4af4229e02bd2a902c6817b423d3b6b6335e31d30e514cba3e7b89ebde2a5e3R1-R53) [[2]](diffhunk://#diff-e7c1fc5465cde83154a81f98052c4c7f1ab71bec52fb2af8670a07896f0d0b76R1-R53) [[3]](diffhunk://#diff-da2ad8c2c1161440c74665181315386b8d1fcc21fbb3b5fe88175cf4a17383c2R1-R79) [[4]](diffhunk://#diff-b075eb823a31ee232ceacd9b5667ab59a4d81082c6a87b7063bcbdebb00ca721R1-R62)
* Modularizes the new core logic and tests under the `core` module, updating `mod.rs` and main module imports for integration. [[1]](diffhunk://#diff-7f67b2dbac45d35ca025e1037b855083c7ca85a8b6be03ee24243aa4681af496R1-R3) [[2]](diffhunk://#diff-821c532dc48536f9fbdd4ddec22629d14b54612d5dd4819b5ebb9cc71dad5da1R1-R4) [[3]](diffhunk://#diff-01fd5002f7585d6f2b853a8c83e6c0a1acb2b16b6e6f8f581fb3365377d331bcR4)

### CODE REVIEW

File: server/src/core/balances_map.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        new_balances.entry(user_id).and_modify(|previous_balance| {
+            *previous_balance += balance.clone();
+        });
+
+        let new_group_balance = self.group_balance.clone() + balance;
+
+        // Retornamos con Ok()
+        Ok(BalancesMap::new(new_balances, new_group_balance))
```
[the remark this answers is not in the document]
Comment (MateoDi9z) [reply server/src/core/balances_map.rs:38]: por ahora safa

File: server/src/core/balances_map.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            return Ok(BalancesMap::new(
+                self.balances.clone(),
+                self.group_balance.clone(),
+            ));
+        }
+        //TODO CREAR ESTRUCTURA QUE SE ENCARGUE DE ESTO Y MANTENGA LOGICA DE DECIMALES ESTILO ETH
+
+        // 3. Configuración de precisión (18 decimales)
+        let escala = 18;
+        let unidad_minima = BigDecimal::from_str("0.000000000000000001").unwrap();
```
[the remark this answers is not in the document]
Comment (MateoDi9z) [reply server/src/core/balances_map.rs:72]: No va a cambiar tampoco

File: server/src/core/balances_map.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Facugarciar0805) [reply server/src/core/balances_map.rs:38]: si por ahi nos pasamos con lo de que todo sea inmutable por ahi convenia con un mapa normal y listo cosa de evitar los clone()

File: server/src/core/core.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            MyTransactionType::Deposit => balances
+                .add_balance_to_user(tx.user_id, tx.amount)
+                .expect("Error de Integridad: Usuario no encontrado."),
+            MyTransactionType::Withdraw => balances
+                .remove_to_all(tx.amount)
+                .expect("Error de Integridad: Fondos insuficientes."),
+            MyTransactionType::Investment => {
+                todo!()
+            }
+            _ => balances,
```
[the remark this answers is not in the document]
Comment (MateoDi9z) [reply server/src/core/core.rs:46]: banco mas el panic, pero si

File: server/src/core/core.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    for expense in expenses {
+        balances = balances
+            .add_balance_to_user(expense.user_id, expense.amount.clone())
+            .expect("Error de Integridad: Usuario de expense no encontrado.");
+
+        balances = balances
+            .remove_to_all(expense.amount)
+            .expect("Error de Integridad: Fondos insuficientes en expense.");
+    }
+    balances
```
[the remark this answers is not in the document]
Comment (MateoDi9z) [reply server/src/core/core.rs]: uuhh si hay que usar el result, asi podemos patear y mapear el error si pincha

File: server/src/core/mod.rs [hunk 5]
```diff
@@ -0,0 +1,3 @@
+pub mod balances_map;
+pub mod core;
```
[the remark this answers is not in the document]
Comment (MateoDi9z) [reply server/src/core/mod.rs:2]: esto estaria bueno cambiarlo

File: server/src/core/balances_map.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MateoDi9z) [reply server/src/core/balances_map.rs:38]: si puede ser

File: server/src/core/core.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Facugarciar0805) [reply server/src/core/core.rs]: esto es el handeleo de errores, por ahora lo dejamos asi dsp hjay que hacer unos lindos custom 
CoreError y manejarnos con eso nomas

**Review by MateoDi9z — COMMENTED** [review]
Excelente!

**Review by Facugarciar0805 — APPROVED** [review]
